File: Persistent/Strings/reverseString.cpp
#include <bits/stdc++.h>

using namespace std;

string reverseStr(string key)
{
  int left = 0, right = key.size() - 1;
  while (left < right)
  {
    char temp = key[left];
    key[left] = key[right];
    key[right] = temp;
    left ++;
    right --;
  }

  return key;
}

string reverseStr2(string key)
{
  stack<char> chars;
  string result = "";

  for (auto& it : key)
  {
    chars.push(it);
  }

  while (!chars.empty())
  {
    char top = chars.top();
    result.push_back(top);
    chars.pop();
  }

  return result;
}

int main()
{
  string key = "I love apples";
  cout << reverseStr2(key);
}

File: Persistent/Strings/reverseString.py
key = "I love apples"

def reverseStr(key):
  left, right = 0, len(key) - 1
  keyArr = list(key)
  while (left < right):
    keyArr[left], keyArr[right] = keyArr[right], keyArr[left]
    left += 1
    right -= 1
  
  return "".join(keyArr)

def reverseStr2(key):
  stack = []
  result = ""
  for i in range(len(key)):
    stack.append(key[i])
  
  for i in range(len(stack)):
    result += stack.pop()

  return result


print(reverseStr2(key))
